packages/opentelemetry-core/src/common/attributes.ts

```diff
@@ -47,6 +47,65 @@ export function isAttributeValue(val: unknown): val is AttributeValue {
   return isValidPrimitiveAttributeValue(val);
 }
 
+export function truncateValueIfTooLong(
+  value: AttributeValue,
+  limit: number | undefined,
+  logger: Logger
+): AttributeValue {
+  if (limit == null) {
+    return value;
+  }
+
+  if (typeof value === 'boolean' || typeof value === 'number') {
+    // these types reasonably can't exceed a limit
+    // if the limit was set to e.g. 2, then true with a length of 4 will exceed a limit
+    // however, the purpose of the limit is to prevent accidents such as values being thousands of characters long
+    return value;
+  }
+
+  if (Array.isArray(value)) {
+    let accruedLength = 0;
+
+    // roughly measure the size of a serialized array
+    // this is done to avoid running json stringify each time attributes are added
+    for (const element of value) {
+      if (element == undefined) {
+        accruedLength += 4;
+      } else if (typeof element === 'string') {
+        accruedLength += element.length;
+      } else {
+        accruedLength += element.toString().length;
+      }
+
+      if (accruedLength > limit) {
+        break;
+      }
+    }
+
+    if (accruedLength > limit) {
+      return truncateValueIfTooLong(JSON.stringify(value), limit, logger);
+    } else {
+      return value;
+    }
+  }
+
+  if (typeof value !== 'string') {
+    logger.warn(
+      `truncateValueIfTooLong expected a value of type string, but received ${typeof value}.`
+    );
+
+    // while this is an undesireable situation, if a limit was set, then our goal is to prevent large payloads
+    return truncateValueIfTooLong(JSON.stringify(value), limit, logger);
+  }
+
+  if (value.length > limit) {
+    logger.warn('Value was truncated.');
```

Comment:
according to spec 
```
To prevent excessive logging, the log should not be emitted more than once per span, attribute, event, or link.
```

here it will be emitted everytime

Reply:
I've adjusted the spec to say "the log MUST NOT be emitted more than once per item on which an attribute is set." and also added a check to only warn once per `Span`.